### PR TITLE
[feature fix] Fix component title length

### PR DESCRIPTION
**Pull request by geeksnglitter**
Purpose
-------------------------
Currently when making a new component on the osf a modal will pop up and allow you to enter a title as long as you wish, but if your title is longer than 200 characters, the component will not be created.

Closes Issue https://github.com/CenterForOpenScience/openscienceframework.org/issues/2993

Changes
------------------
I simply added a maxlength="200" into the field for component titles. Stopping users form being able to enter a title longer than 200 characters.

Edit: Because we want there to be consistency in title length, there were many changes in order to secure that title length remains around 200 throughout the system. Both front end and back end validations were added in order to enforce the 200 character limit on titles. You can no longer edit a title to be as long as you desire as shown in https://github.com/CenterForOpenScience/osf.io/issues/2993#issuecomment-110046427.

Projects will now not be able to be created if a user attempts to make a title past 200 characters. They will also not be able to enter titles past 200 characters in the fields that edit said titles, so project creation should not fail for this reason.

### CODE REVIEW

**Comment by sloria** [comment]
As mentioned in https://github.com/CenterForOpenScience/osf.io/issues/2993#issuecomment-110058487, validation should also happen on the backend.

File: website/static/js/nodeControl.js [hunk 1]
```diff
@@ -185,16 +185,20 @@ var ProjectViewModel = function(data) {
...
+        $('#nodeTitleEditable').editable($.extend({},editableOptions, {
             name: 'title',
             title: 'Edit Title',
+            tpl: "<input type='text' maxlength='200'>",
             validate: function (value) {
                 if ($.trim(value) === '') {
                     return 'Title cannot be blank.';
                 }
+                else if(value.length > 200){
+                    return 'Titles cannot be longer than 200 characters.';
```
Comment (samanehsan) [inline website/static/js/nodeControl.js]: Please change this to "Title" instead of "Titles"

File: website/project/model.py [hunk 2]
```diff
@@ -507,13 +506,16 @@ def validate_category(value):
...
 
 def validate_title(value):
-    """Validator for Node#title. Makes sure that the value exists.
+    """Validator for Node#title. Makes sure that the value exists and is not
+    above 200 characters.
     """
     if value is None or not value.strip():
         raise ValidationValueError('Title cannot be blank.')
+    elif len(value) > 200:
+            raise ValidationValueError('Titles cannot be longer than 200 ' +
```
Comment (samanehsan) [inline website/project/model.py]: Please change this to "Title" instead of "Titles" as well.

File: website/static/js/project.js [hunk 3]
```diff
@@ -136,7 +136,7 @@ $(function() {
 
             e.preventDefault();
         } else if ($(e.target).find('#title').val().length > 200) {
-            $('#alert').text('The new component title cannot be more than 200 characters.');
+            $('#alert').text('The new component title cannot be more than 200 characters.'); //This alert never appears...
```
Comment (samanehsan) [inline website/static/js/project.js]: Can you add a div with `id=alert` (and the `text-danger` css class for styling) to `modal_add_component.mako`? It must have been deleted accidentally, because right now even the alert for an empty component title won't show.

Comment (samanehsan) [inline website/static/js/project.js]: Do we need to include an error message if it's not possible for the user to enter a title longer than 200 characters? @sloria

File: website/project/model.py [hunk 4]
```diff
@@ -1451,8 +1453,9 @@ def set_title(self, title, auth, save=False):
         :param str title: The new title.
         :param auth: All the auth information including user, API key.
         """
-        if title is None or not title.strip():
-            raise ValidationValueError('Title cannot be blank.')
+        #Called so validation does not have to wait until save.
+        validate_title(title)
```
Comment (samanehsan) [inline website/project/model.py]: Why validate the title prior to saving?

File: website/project/views/node.py [hunk 5]
```diff
@@ -143,9 +143,6 @@ def folder_new_post(auth, node, **kwargs):
 
     title = request.json.get('title')
 
-    if not title or len(title) > 200:
```
Comment (samanehsan) [inline website/project/views/node.py]: Why was this removed?

**Comment by samanehsan** [comment]
Pass complete! :sunglasses:

File: website/static/js/project.js (hunk 3 above)
Comment (geeksnglitter) [inline website/static/js/project.js]: Ya I didn't write that. It never did anything as far as I can tell.

File: website/project/views/node.py (hunk 5 above)
Comment (geeksnglitter) [inline website/project/views/node.py]: There were inconsistant/ random validations like throughout the code. Because this saves at the end, it runs validate_title and checks those two things already. Having that line would be redundant.

File: website/project/model.py (hunk 4 above)
Comment (geeksnglitter) [inline website/project/model.py]: Otherwise I'd have to add my try/except every time I save a node, instead of every time I just change a title. This allows us to find the validation errors earlier and facilitates error handling and testing.

Comment (samanehsan) [inline website/project/model.py]: :+1:

File: website/project/views/node.py (hunk 5 above)
Comment (samanehsan) [inline website/project/views/node.py]: Got it.

**Comment by samanehsan** [comment]
@sloria this is ready for another pass! :palm_tree:

File: tests/test_models.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 import datetime
 import urlparse
 import itsdangerous
+import random, string
```
Comment (lyndsysimon) [inline tests/test_models.py]: Please break this into two lines.

ref: https://www.python.org/dev/peps/pep-0008/#imports

File: website/project/model.py [hunk 7]
```diff
@@ -510,13 +509,15 @@ def validate_category(value):
 
 
 def validate_title(value):
-    """Validator for Node#title. Makes sure that the value exists.
+    """Validator for Node#title. Makes sure that the value exists and is not
+    above 200 characters.
     """
     if value is None or not value.strip():
         raise ValidationValueError('Title cannot be blank.')
+    elif len(value) > 200:
```
Comment (lyndsysimon) [inline website/project/model.py]: `if` and `elif` are identical here, since the prior block raises an exception - but for clarity, it's generally best to use `if` and space them out like so:

```python

if value is None or not value.strip():
    raise ValidationValueError('Title cannot be blank.')

if len(value) > 200:
    raise ValidationValueError('Title cannot exceed 200 characters.')

return True
```

No need to change this, I'm just being me.

File: website/templates/project/modal_add_component.mako [hunk 8]
```diff
@@ -9,7 +9,8 @@
                 </div><!-- end modal-header -->
                 <div class="modal-body">
                     <div class="form-group">
-                        <input id="title" placeholder="Component Title" name="title" type="text" class='form-control'>
+                        <input id="title" maxlength="200" placeholder="Component Title" name="title"  type="text" class='form-control'>
+                        <div id="alert"></div>
```
Comment (lyndsysimon) [inline website/templates/project/modal_add_component.mako]: Do we have to use the ID "alert" here? It seems like that's a value that would be very likely to be used elsewhere.

I propose making it a class instead (`modal-alert`), and using a compound selector in `project.js` to grab it (`#newComponent .modal-alert`).

**Comment by lyndsysimon** [comment]
Pass complete. :cake:
